src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.android.cs

```diff
@@ -32,39 +32,99 @@ public double Balance
+	bool isPlaying = false;
+
+
+	public void SetSpeed(double sp)
 	{
-		get => player.PlaybackParams.Speed;
-		set
+		if (!OperatingSystem.IsAndroidVersionAtLeast(23))
+		{
+			System.Diagnostics.Trace.TraceWarning("Setting speed is only supported on Android 23 and above.");
+			return;
+		}
+
+		try
 		{
-			// Check if set speed is supported
-			if (CanSetSpeed)
+			if (isChangingSpeed)
 			{
-				// Speed on Android can be between 0 and 6
-				var speedValue = Math.Clamp((float)value, 0.0f, 6.0f);
+				System.Diagnostics.Trace.TraceWarning("last speed update was not yet completed");
+				return;
+			}
 
-				if (float.IsNaN(speedValue))
-				{
-					speedValue = 1.0f;
-				}
+			isChangingSpeed = true;
 
-				player.PlaybackParams = player.PlaybackParams.SetSpeed(speedValue) ?? player.PlaybackParams;
-			}
-			else
+			// internal (android) speed value (float)
+			internalSpeed = Math.Clamp((float)sp, minSpeed, maxSpeed);
+
+			// shared speed value (double)
+			speed = Math.Clamp(sp, MinimumSpeed, MaximumSpeed);
+
+			// to prevent illegal exception when changing the speed on some devices,
+			// we need to change the speed while the audio is reset, see: https://stackoverflow.com/questions/39442522/setplaybackparams-causes-illegalstateexception
+
+			// store current state before resetting
+			var previousPosition = stopwatch.ElapsedMilliseconds;
+			var wasPlaying = isPlaying;
+
+			// reset
+			isPlaying = false;
+			player.Reset();
+
+			// after reset, we need to prepare the audio source again
+			PrepareAudioSource();
+
+			// now we are going to update the speed parameter of the actual audio player
+
+			// allow defaults ensures that we can play the audio in case the pitch value is not set. When the audioplayer is paused, PlaybackParams may not be what we would like it to be
+			var parms = (player.PlaybackParams.AllowDefaults() ?? new PlaybackParams().AllowDefaults())?.SetSpeed(internalSpeed) ?? throw new ArgumentException("speed value not supported");
+
+			// this ensures that the audio will fail if the speed is not supported, otherwise android simply continues but does not play any audio
+			// https://developer.android.com/reference/android/media/PlaybackParams
+			parms.SetAudioFallbackMode((int)AudioFallbackMode.Fail);
+
+			player.PlaybackParams = parms;
+
+			// we need to update the stopwatch to reflect the new speed
+			stopwatch = new AudioStopwatch(TimeSpan.FromMilliseconds(previousPosition), speed);
+
+			// because we had to reset, we now restore the previous state of the audio player
+			player.SeekTo((int)previousPosition);
+			if (wasPlaying)
 			{
-				throw new NotSupportedException("Set playback speed is not supported!");
+				isPlaying = true;
+				stopwatch.Start();
+				player.Start();
 			}
+
+			isChangingSpeed = false;
 		}
+		finally
+		{
+			isChangingSpeed = false;
+		}		
 	}
 
+
+
+	float internalSpeed = 1.0f;
+	double speed = 1.0;
+	public double Speed => speed;
+
+	const float minSpeed = 0;
```

Comment:
Why do we have constants and a property? If they are both needed the property should return the constant

Reply:
One is public API and the other is for internal calculations, but I agree, there is probably a more efficient way. Feel free to change it.

Reply:
Oh it's doubles vs floats. My bad